structure/redblacktree_test/rbtree_test.go

```diff
@@ -0,0 +1,108 @@
+package redblacktree_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/TheAlgorithms/Go/structure/redblacktree"
+)
+
+type RBTree = redblacktree.RBTree
+type RBNode = redblacktree.RBNode
+
+func compare(tree *RBTree, preVal []int, preRed []bool, inVal []int, inRed []bool) bool {
+	// fmt.Println("In compare")
+	var ret = false
+	var pre = redblacktree.PreOrder(tree.Root)
+	var in = redblacktree.InOrder(tree.Root)
+	if len(pre) != len(preVal) || len(in) != len(inVal) {
+		return ret
+	}
+	for i := 0; i < len(pre); i++ {
+		if !pre[i].EquivalentNode(redblacktree.CreateNode(preVal[i], preRed[i])) {
+			fmt.Printf("Error at pre %d", i)
+			return ret
+		}
+	}
+	for i := 0; i < len(in); i++ {
+		if !in[i].EquivalentNode(redblacktree.CreateNode(inVal[i], inRed[i])) {
+			return ret
+		}
+	}
+	return true
+}
+func TestInsert(t *testing.T) {
+	var tree = &RBTree{}
+	var vals = []int{8, 18, 5, 15, 17, 25, 40, 80}
+	var expPreVal = []int{17, 8, 5, 15, 25, 18, 40, 80}
+	var expPreRed = []bool{false, true, false, false, true, false, false, true}
+	var expInVal = []int{5, 8, 15, 17, 18, 25, 40, 80}
+	var expInRed = []bool{false, true, false, false, false, true, false, true}
```

Comment:
This is why I don't like this implementation. You have to keep track of more than 5 different indices just to compare the result!!

Reply:
You have two data structures `RBNode` and `RBTree` but there is no need for `RBTree` right? You are not using root in any way which doesn't directly use the `RBNode`. In fact, you are just using `RBNode` in everything defined for the tree. In this implementation, the responsibility of RBTree struct depends directly on the implementation details of RBNode. The first principle in SOLID design pattern is to avoid such things. You can ask yourself a question, do you really need `RBTree`? In every method that you define for `RBTree` you are using the fields of `RBNode` continuously because these two are so linked together that you cannot decouple them - not good if you are trying to make production applications. Note that trees by nature are naturally recursive in nature but because of your `RBTree` implementation, you made this very difficult to do recursion on it. Anyhow, I recommend you to look at the implementation of Red Black Tree in our python repo - [here](https://github.com/TheAlgorithms/Python/blob/master/data_structures/binary_tree/red_black_tree.py)

Anyhow this is my suggestion as I recognise that you have spend quite a bit of time with this and if you think that my suggestion here is unreasonable, just comment here. I can then ask the other maintainers for their opinions on this and since it seems that they had previously approved it, they might consider your implementation to be fine if they think it can be merged, then I will agree with them. Hope these suggestions help.